src/components/instructor/AddUsersFromFile.vue

```diff
@@ -1,23 +1,325 @@
+		const selected: Ref<Array<Dictionary<string|number>>> = ref([]); // stores the selected users
+		const column_headers: Ref<Dictionary<string>> = ref({});
+		const user_param_map: Ref<Dictionary<string>> = ref({}); // provides a map from column number to field name
+
+		const first_row_header: Ref<boolean> = ref(false);
+		const header_row: Ref<Dictionary<string|number>> = ref({});
+		const users_to_add: Ref<Array<User>> = ref([]);
+		const course_users_to_add: Ref<Array<Dictionary<string|number>>> = ref([]);
+		// const course_users_to_add: Ref<Array<CourseUser>> = ref([]);
+
+		const user_fields = [
+			{ label: 'Username', field: 'username', regexp: /(user)|(login)/i },
+			{ label: 'First Name', field: 'first_name', regexp: /first/i },
+			{ label: 'Last Name', field: 'last_name', regexp: /last/i },
+			{ label: 'Student ID', field: 'student_id', regexp: /stud/i },
+			{ label: 'Email', field: 'email', regexp: /email/i },
+			{ label: 'Section', field: 'section', regexp: /sect/i },
+			{ label: 'Recitation', field: 'recitation', regexp: /recit/i },
+			{ label: 'Role', field: 'role', regexp: /role/i } ];
+
+		// use the first row to find the headers of the table (field names)
+		// this is based on trying to match the needed field names to the know user
+		// field names.
+		const fillHeaders = () => {
+			Object.keys(header_row.value).forEach((key) => {
+				user_fields.forEach((field) => {
+					if (field.regexp.test(`${header_row.value[key]}`)) {
+						column_headers.value[key] = field.label;
+					}
+				});
+			});
+		};
+
+		watch([column_headers], () => {
+
+			// these are the defined columns in the table
+			const def_cols = pick(column_headers.value, Object.keys(column_headers.value));
+
+			// this is an object of the fields selected from the table
+			const fields = pick(fromPairs(user_fields.map((obj) => [obj.label, obj.field])), values(def_cols));
+			// mapping of column number to user field
+			user_param_map.value =  invert(mapValues(def_cols, (obj)=>fields[obj]));
+		},
+		{ deep: true });
+
+		watch([first_row_header], () => {
+			if(first_row_header.value) {
+				const first_row = users.value.shift();
+				if (first_row) {
+					header_row.value = first_row;
+					fillHeaders();
+				}
+			} else {
+				users.value.unshift(header_row.value);
+			}
+		});
+
+		watch([selected], () => {
+			validated.value = true;
+			users_to_add.value = [];
+			course_users_to_add.value = [];
+			invalid_table_cells.value = [];
+			selected.value.forEach((params) => {
+
+				let user: Dictionary<string|number> = {};
+				let course_user: Dictionary<string|number> = {};
+
+				try {
+					user = pick(mapValues(user_param_map.value, (obj) => params[obj]), Object.keys(newUser()));
+					users_to_add.value.push(parseUser(user));
+
+					course_user = pick(mapValues(user_param_map.value, (obj) => params[obj]),
+						Object.keys(newCourseUser()));
+
+					const parsed_course_user = parseCourseUser(course_user);
```

Comment:
before parsing `course_user`, we need `course_user.role = course_user.role ?? 'student'` or else `parseCourseUser` will fail.

Alternatively, `parseCourseUser` should not throw an error with a missing user.role, and instead provide 'student' as a default.

Reply:
Agree with this.  I've been thinking of a better way to handle all of the parsing.